core/spec/models/spree/shipment_spec.rb

```diff
@@ -592,14 +528,59 @@
     end
   end
 
-  context "#tracking_url" do
-    subject do
-      shipment.tracking_url
+  describe "#ship" do
+    context "when the shipment is canceled" do
+      let(:address){ build(:address) }
+      let(:order){ create(:order_with_line_items, ship_address: address) }
+      let(:shipment_with_inventory_units) { create(:shipment, order: order, address: address, state: "canceled") }
+
+      subject { shipment_with_inventory_units.ship! }
+
+      it "unstocks them items" do
+        allow(shipment_with_inventory_units.stock_location).to receive(:unstock).with(an_instance_of(Spree::Variant), 1, shipment_with_inventory_units)
+        subject
+      end
+    end
+
+    ["ready", "canceled"].each do |state|
+      context "from #{state}" do
+        before do
+          allow(order).to receive(:update!)
+          allow(shipment).to receive_messages(require_inventory: false, update_order: true, state: state)
+        end
+
+        it "should call fulfill_order_with_stock_location" do
+          expect(Spree::OrderStockLocation).to(
+            receive(:fulfill_for_order_with_stock_location).
+            with(order, stock_location)
+          )
+          shipment.ship!
+        end
+
+        it "finalizes adjustments" do
+          shipment.adjustments.each do |adjustment|
+            expect(adjustment).to receive(:finalize!)
+          end
+          shipment.ship!
+        end
+      end
+    end
+  end
+
+  describe "#ready" do
+    # Regression test for #2040
```

Comment:
@alexblackie @jhawthorn One regression test was removed earlier in Alex' work (https://github.com/solidusio/solidus/pull/433/files#diff-079dc32783aecdf3902b9cdb9f2bfd41L28); should this be removed? Another one https://github.com/solidusio/solidus/pull/433/files#diff-079dc32783aecdf3902b9cdb9f2bfd41R606.